### PR TITLE
Initial implementation of Inheritance and Polymorphic functions

### PR DESCRIPTION
**Pull request by tanay-man**

### CODE REVIEW

**Review by certik — APPROVED** [review]
I think this is good. @Thirumalai-Shaktivel if you could also review it, that would be great.

**Comment by certik** [comment]
Not sure why the kernel is failing:
```
[doctest] doctest version is "2.4.8"
[doctest] run with "--help" for options
0 0 0 0 0 0 0 0 0 0 
test_lpython: /home/runner/micromamba/envs/lp/include/llvm/IR/DataLayout.h:636: uint64_t llvm::StructLayout::getElementOffset(unsigned int) const: Assertion `Idx < NumElements && "Invalid element idx!"' failed.
===============================================================================
/home/runner/work/lpython/lpython/src/lpython/tests/test_llvm.cpp:1556:
TEST CASE:  PythonCompiler classes

/home/runner/work/lpython/lpython/src/lpython/tests/test_llvm.cpp:1556: FATAL ERROR: test case CRASHED: SIGABRT - Abort (abnormal termination) signal

```

It looks like it could be caused by this PR.

File: integration_tests/class_05.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+from lpython import i32
+
+class Animal:
+    def __init__(self:"Animal"):
+        self.species: str = "Generic Animal"
+        self.age: i32 = 0
+
+class Dog(Animal):
+    def __init__(self:"Dog", name:str, age:i32):
+        # super().__init__()
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/class_05.py:10]: TODO

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 2]
```diff
@@ -3327,6 +3335,15 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
...
                 current_scope = al.make_new<SymbolTable>(parent_scope);
+                // if ( parent ) {
+                //     ASR::Struct_t* base_st = ASR::down_cast<ASR::Struct_t>(parent);
+                //     SymbolTable* base_scope = base_st->m_symtab;
+                //     for (auto i : base_scope->scope) {
+                //         std::string name = i.first;
+                //         ASR::symbol_t* sym = i.second;
+                //         current_scope->add_symbol(name,sym);
+                //     }
+                // }
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/semantics/python_ast_to_asr.cpp:3346]:
```suggestion
```

**Review by Thirumalai-Shaktivel — APPROVED** [review]

**Comment by Thirumalai-Shaktivel** [comment]
I think the kernel failure is not related to this PR.

**Comment by tanay-man** [comment]
We have to merge this with https://github.com/lcompilers/lpython/pull/2795 before merging as some of the ASR-pass related modifications are present there.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 3]
```diff
@@ -8100,7 +8156,19 @@ we will have to use something else.
                         for (size_t i=0; i<args.n; i++) {
                             new_args.push_back(al, args[i]);
                         }
-                        st = get_struct_member(der, call_name, loc);
+                        if ( der->m_symtab->get_symbol(call_name) ) {
+                            st = get_struct_member(der_sym, call_name, loc);
+                        } else if ( der->m_parent ) {
+                            ASR::Struct_t* parent = ASR::down_cast<ASR::Struct_t>(der->m_parent);
+                            if ( !parent->m_symtab->get_symbol(call_name) ) {
+                                throw SemanticError("Method not found in the class or parents",loc);
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/semantics/python_ast_to_asr.cpp:8164]:
```suggestion
                                throw SemanticError("Method not found in the class "+ std::string(der->m_name) +
                                    " or it's parents",loc);
```

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 4]
```diff
@@ -5166,31 +5190,82 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+                        ASR::symbol_t* call_sym = get_struct_member(parent_sym,call_name,loc);
+                        super_call_stmt = ASRUtils::STMT(
+                            ASR::make_SubroutineCall_t(al, loc, call_sym, call_sym, args_w_first.p, 
+                                                        args_w_first.size(), nullptr));
+                }
+            } else {
+                body.push_back(al, x.m_body[i]);
             }
         }
         current_scope = f->m_symtab;
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/semantics/python_ast_to_asr.cpp:5251]: Seems to be repeated
```suggestion
```

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 5]
```diff
@@ -5166,31 +5190,82 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
-        Vec<ASR::stmt_t*> body;
-        body.reserve(al, x.n_body);
+        Vec<ASR::stmt_t*> body_asr;
+        body_asr.reserve(al, x.n_body);
+        Vec<ASR::stmt_t*> new_body_asr;
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/semantics/python_ast_to_asr.cpp:5253]: Why not push to the body_asr only?

**Review by Thirumalai-Shaktivel — COMMENTED** [review]
Thanks for working on this. It looks great!

I left minor suggestions

**Comment by Thirumalai-Shaktivel** [comment]
Update or rebase this PR and it is good to merge.

File: src/libasr/pass/pass_utils.cpp [hunk 6]
```diff
@@ -1281,7 +1281,6 @@ namespace LCompilers {
         }
         return import_struct_member;
     }
-
-    } // namespace PassUtils
+} // namespace PassUtils
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/pass_utils.cpp:1284]:
```suggestion

    } // namespace PassUtils
```

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 5 above)
Comment (Thirumalai-Shaktivel) [reply src/lpython/semantics/python_ast_to_asr.cpp:5254]: Good, resolve this and we can merge.

File: integration_tests/class_05.py (hunk 1 above)
Comment (Thirumalai-Shaktivel) [reply integration_tests/class_05.py]: The changes has been made in this PR itself.

**Comment by Thirumalai-Shaktivel** [comment]
Also, change the PR heading to appropriate description

**Comment by Thirumalai-Shaktivel** [comment]
Also, let's clean up the commit messages.

**Comment by Thirumalai-Shaktivel** [comment]
See: https://github.com/lcompilers/lpython/pull/783#issuecomment-1188875210

**Comment by tanay-man** [comment]
Do I have to remove all the commits in the history?

**Comment by tanay-man** [comment]
@Thirumalai-Shaktivel  can you please tell what an appropriate title for this PR would be?

**Comment by tanay-man** [comment]
fixes #2799 #2797

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 5 above)
Comment (Thirumalai-Shaktivel) [reply src/lpython/semantics/python_ast_to_asr.cpp:5254]: I think new_body_asr seems to be redundant. Can you please check?

**Comment by Thirumalai-Shaktivel** [comment]
Thanks for the changes. I have set this PR to auto-merge

**Pull request merged**